### PR TITLE
executor: also collect unchanged unique keys for lock

### LINKED ISSUE #36438: Unique key lock not consistent when row is changed or not
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
mysql> show create table t;
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                         |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t     | CREATE TABLE `t` (
  `i` varchar(20) NOT NULL,
  PRIMARY KEY (`i`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

mysql> insert into t values ('a');
Query OK, 0 rows affected (0.00 sec)
Rows matched: 1  Changed: 0  Warnings: 0

mysql> begin;
Query OK, 0 rows affected (0.00 sec)

mysql> select * from t;
+---+
| i |
+---+
| a |
+---+
1 row in set (0.00 sec)

mysql> update t set i = 'a'; -- insert into t values ('a') in another session
Query OK, 0 rows affected (0.00 sec)
Rows matched: 1  Changed: 0  Warnings: 0

mysql> update t set i = 'b'; -- insert into t values ('b') in another session
Query OK, 1 row affected (0.00 sec)
Rows matched: 1  Changed: 1  Warnings: 0

mysql> rollback;
Query OK, 0 rows affected (0.00 sec)
```

### 2. What did you expect to see? (Required)
The first insert in another session should wait lock.

### 3. What did you see instead (Required)
ERROR 1062 (23000): Duplicate entry 'a' for key 'PRIMARY'

### 4. What is your TiDB version? (Required)

01d02568c25946684b6c4cc4cc5343f6d2068a99

### PR DESCRIPTION
**Pull request by zyguan**
Signed-off-by: zyguan <zhongyangguan@gmail.com>

### What problem does this PR solve?

Issue Number: close #36438

Problem Summary: Only row keys will be locked after update nothing, which leads to #36438.

### What is changed and how it works?

Also collect unchanged unique keys and lock them.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by jackysp — APPROVED** [review]
LGTM

File: executor/write.go [hunk 1]
```diff
@@ -158,7 +158,21 @@ func updateRecord(ctx context.Context, sctx sessionctx.Context, h kv.Handle, old
 		unchangedRowKey := tablecodec.EncodeRowKeyWithHandle(physicalID, h)
 		txnCtx := sctx.GetSessionVars().TxnCtx
 		if txnCtx.IsPessimistic {
-			txnCtx.AddUnchangedRowKey(unchangedRowKey)
+			txnCtx.AddUnchangedLockKey(unchangedRowKey)
+			for _, idx := range t.Indices() {
+				if !idx.Meta().Unique {
+					continue
+				}
```
Comment (tangenta) [inline executor/write.go:165]: I tested with MySQL 8.0.27. It seems that the non-unique keys are locked as well:

```sql
drop table if exists t;
create table t (a varchar(10), b varchar(10), unique key(a), key(b));
insert into t values ('a0', 'b0');
begin;
update t set a = 'a0' where b = 'b0'; -- insert into t values ('a1', 'b0'); <- Blocked in another session
rollback;
```

I am not sure which behavior is reasonable.

Comment (jackysp) [reply executor/write.go:165]: I think we could leave the locking for the non-unique index to other places. It should be considered with changed/non-changed rows together.

**Review by tangenta — APPROVED** [review]

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**